c2rust-transpile/src/c_ast/mod.rs

```diff
@@ -486,6 +492,71 @@ impl TypedAstContext {
         self.index(resolved_typ_id)
     }
 
+    /// Extract decl of referenced function.
+    /// Looks for ImplicitCast(FunctionToPointerDecay, DeclRef(function_decl))
+    pub fn function_declref_decl(&self, func_expr: CExprId) -> Option<&CDeclKind> {
+        use CastKind::FunctionToPointerDecay;
+        if let CExprKind::ImplicitCast(_, fexp, FunctionToPointerDecay, _, _) = self[func_expr].kind
+        {
+            if let CExprKind::DeclRef(_ty, decl_id, _rv) = &self[fexp].kind {
+                let decl = &self.index(*decl_id).kind;
+                assert!(matches!(decl, CDeclKind::Function { .. }));
+                return Some(decl);
+            }
+        }
+        None
+    }
+
+    /// Return the list of types for a list of declared function parameters.
+    pub fn tys_of_params(&self, parameters: &[CDeclId]) -> Option<Vec<CQualTypeId>> {
+        let mut param_tys = vec![];
+        for p in parameters {
+            match self.index(*p).kind {
+                CDeclKind::Variable { typ, .. } => param_tys.push(CQualTypeId::new(typ.ctype)),
+                _ => return None,
+            }
+        }
+        return Some(param_tys);
+    }
+
+    /// Return the most precise possible CTypeKind for the given function declaration.
+    /// Specifically, ensures that arguments' types are not resolved to underlying types if they were
+    /// declared as typedefs, but returned as those typedefs.
+    ///
+    /// The passed CDeclId must refer to a function declaration.
+    pub fn fn_decl_ty_with_declared_args(&self, func_decl: &CDeclKind) -> CTypeKind {
+        if let CDeclKind::Function {
+            typ, parameters, ..
+        } = func_decl
+        {
+            let typ = self.resolve_type_id(*typ);
+            let decl_arg_tys = self.tys_of_params(parameters).unwrap();
+            let typ_kind = &self[typ].kind;
+            if let &CTypeKind::Function(ret, ref _arg_tys, a, b, c) = typ_kind {
+                return CTypeKind::Function(ret, decl_arg_tys, a, b, c);
+            }
+            panic!("expected {typ:?} to be CTypeKind::Function, but it was {typ_kind:?}")
+        }
+        panic!("expected a CDeclKind::Function, but passed {func_decl:?}")
+    }
+
+    /// Return the id of the most precise possible type for the function referenced by the given
+    /// expression.
+    pub fn function_declref_ty_with_declared_args(
+        &self,
+        func_expr: CExprId,
+    ) -> Option<CQualTypeId> {
+        if let Some(func_decl @ CDeclKind::Function { .. }) = self.function_declref_decl(func_expr)
```

Comment:
`fn function_declref_decl` already `assert!`s that this is a `CDeclKind::Function`.  We don't have variant types in Rust, but we shouldn't need the `None` branch here, which is just confusing if it's unreachable.  Cleanest would be to have a `FunctionDecl` type in `CDeclKind::Function` instead of the fields directly, but I'm not sure how much churn that would cause elsewhere.  Maybe just an `assert!` or `unreachable!` suffices here.

Reply:
No, that `assert!` is just about the correspondence between `FunctionToPointerDecay` and `CDeclKind::Function`. If the expression is not a reference to a function (it's a different kind of expression, or a reference to a non-function decl) then `fn_declref_decl` happily returns `None`. These are intended to be partial mappings that don't panic on expected inputs; the panics are for violations of basic AST invariants.